### PR TITLE
Protect atomic actions

### PR DESCRIPTION
**Pull request by erights**
An atomic action is some set of operations, where either all or none of them should happen. Our style is to test everything that could cause any of the individual operations before the commit point, and then do all the actions, now that it is impossible for any of these actions to fail. However, we might have missed a reason why it might fail anyway. This PR puts `try`/`catch` blocks around the atomic actions we've identified. If the try block exits by throwing, then the catch block terminates some appropriate unit of computation, like the contract or the vat.

Replaces #3144

### CODE REVIEW

File: packages/ERTP/src/paymentLedger.js
Comment (dtribble) [inline packages/ERTP/src/paymentLedger.js]: this is unnecessary because there's only one operation. This here would be better with a try/catch pattern because existing developers and tools would recognize it

File: packages/ERTP/src/paymentLedger.js
Comment (dtribble) [inline packages/ERTP/src/paymentLedger.js]: I think this woudl be more familiar to programmers as a try/catch pattern:

```
    try {
      // COMMIT POINT
      payments.forEach(payment => paymentLedger.delete(payment));

      const newPayments = newPaymentBalances.map(balance => {
        const newPayment = makePayment();
        paymentLedger.init(newPayment, balance);
        return newPayment;
      });
      return harden(newPayments);
    } catch (ex) {
      throw fatalAssertAfterCommit(ex);
    }
```

Comment (dtribble) [reply packages/ERTP/src/paymentLedger.js]: In ZCF code we have other options that are still less syntactic overhead.

File: packages/zoe/src/contractFacet/zcfSeat.js
Comment (dtribble) [inline packages/zoe/src/contractFacet/zcfSeat.js]: Given the comment above, should this fail the vat later if the replaceAllocations breaks?  It makes me very uncomfortable to see a hanging async operation not get rolled up into "kill vat if this breaks"

**Review by dtribble — CHANGES_REQUESTED** [review]
As in the library version, I find this code harder to read and be sure what's going on.

File: packages/ERTP/src/paymentLedger.js
Comment (erights) [reply packages/ERTP/src/paymentLedger.js]: > In ZCF code we have other options that are still less syntactic overhead.

What do you have in mind?

**Comment by erights** [comment]
See https://github.com/endojs/endo/pull/733#issuecomment-864627282

Will remove the dependence on https://github.com/endojs/endo/pull/733 and try again.

File: packages/ERTP/src/paymentLedger.js
Comment (erights) [reply packages/ERTP/src/paymentLedger.js]: Agreed. Closed https://github.com/endojs/endo/pull/733

File: packages/ERTP/src/paymentLedger.js
Comment (erights) [reply packages/ERTP/src/paymentLedger.js]: > I think this woudl be more familiar to programmers as a try/catch pattern:

Agreed. Closed https://github.com/endojs/endo/pull/733

But still curious about 

> In ZCF code we have other options that are still less syntactic overhead.

File: packages/zoe/src/contractFacet/zcfSeat.js
Comment (erights) [reply packages/zoe/src/contractFacet/zcfSeat.js]: A failed `replaceAllocations` --- as well as several other failures that happen on the zoe side in service of a particular zcf vat --- must be fatal to that zcf vat. This PR is supposed to ensure that too. I'll leave this conversation unresolved until this specific case is dealt with.

**Comment by katelynsills** [comment]
@erights, is this ready for review? If not, no problem.

File: packages/zoe/src/contractFacet/zcfSeat.js
Comment (erights) [reply packages/zoe/src/contractFacet/zcfSeat.js]: In startInstance.js, I did this for `replaceAllocations` specifically, so resolving. But @katelynsills @dtribble , What other zoe-side operations should, if they fail, kill their zcf?

**Comment by erights** [comment]
I still have to write tests before merging. But what's here is now ready for review. Thanks.

File: packages/zoe/src/zoeService/internal-types.js
Comment (erights) [inline packages/zoe/src/zoeService/internal-types.js:85]: These properties were already there. This PR just adds the missing declarations.

**Comment by erights** [comment]
Ready again for review.

> I still have to write tests before merging. But what's here is now ready for review. Thanks.

That's still true. No tests yet.

This version is simplified from the previous one. It uses the existing terminateWithFailure function abstraction, rather than trying to rebuild it around custom asserts (which would eventually have needed https://github.com/endojs/endo/pull/796 )

File: packages/ERTP/src/paymentLedger.js
Comment (katelynsills) [inline packages/ERTP/src/paymentLedger.js]: Will this be captured as a Github issue for later, or is this still to be done in this PR?

Seems like it's easy to destroy the `paymentLedger` by setting the variable to `undefined` or another `weakStore`, but we would also need to destroy purses, which carry their own balances as internal state. And we would have to prevent any new minting of payments.

File: packages/zoe/src/internal-types.js
Comment (katelynsills) [inline packages/zoe/src/internal-types.js]: This is just my preference, but it would be nice to have the shutdownWithFailure type be a named callback itself, so we can add a comment to the type about the severity of effects and use the same type rather than remaking it each time

Comment (katelynsills) [reply packages/zoe/src/internal-types.js]:
```suggestion
 * @param {ShutdownWithFailure} shutdownWithFailure
```

File: packages/zoe/src/zoeService/startInstance.js
Comment (katelynsills) [inline packages/zoe/src/zoeService/startInstance.js]:
```suggestion
    const { adminNode, root } = zoeInstanceStorageManager;
```

File: packages/zoe/src/zoeService/zoeStorageManager.js
Comment (katelynsills) [inline packages/zoe/src/zoeService/zoeStorageManager.js]:
```suggestion
```

**Review by katelynsills — COMMENTED** [review]
Looks good, I have some small suggestions. Large question on the TODO in `destroyLedgerWithFailure`

File: packages/ERTP/src/paymentLedger.js
Comment (erights) [reply packages/ERTP/src/paymentLedger.js]: Good point about the purses. I'd missed that. Given that, I'll file this as an issue rather than address in this PR.

File: packages/zoe/src/zoeService/startInstance.js
Comment (erights) [reply packages/zoe/src/zoeService/startInstance.js]: An `await` ripe for removal! Thanks for catching this.

File: packages/ERTP/src/paymentLedger.js
Comment (erights) [reply packages/ERTP/src/paymentLedger.js]: https://github.com/Agoric/agoric-sdk/issues/3434

**Comment by erights** [comment]
@katelynsills @michaelfig please note the change I just made to the treasury. The new type mismatch alerted me to this, but the code had a worse problem. It was using an untagged template literal to compose with an error, losing the error object and therefore losing the possibility of logging a stack trace.

**Comment by katelynsills** [comment]
Let's make sure to include @Chris-Hibbert on anything related to the treasury. @Chris-Hibbert, see Mark's note above.

File: packages/treasury/src/vaultManager.js
Comment (Chris-Hibbert) [inline packages/treasury/src/vaultManager.js:188]: This looks surprising to me. When I read it, I see an `assert.error()`, which I expect to terminate operation, which is expected to return a value that will be passed to `zcf.shutdownWithFailure()`. Why is the latter able to gain control after the `assert.error()` call?

Comment (erights) [reply packages/treasury/src/vaultManager.js:188]: Hi @Chris-Hibbert , I think you're thinking of `assert.fail` which both creates the error and throws it. `assert.error` has the same inputs and creates the same error, but returns it rather than throwing it.

**Review by Chris-Hibbert — APPROVED** [review]
The treasury changes LGTM. The rest looks fine, but I'll leave it to @katelynsills to approve.

File: packages/ERTP/src/paymentLedger.js
Comment (katelynsills) [inline packages/ERTP/src/paymentLedger.js]: For clarity, I would prefer if we don't call this `destroyLedgerWithFailure` if it does not destroy the ledger yet. (I don't want to confuse our auditors.) Can we call this method something else and move the contents of the comment into issue 3434?

**Review by katelynsills — APPROVED** [review]
Everything looks good, although I have a strong preference about removing mention of destroying the ledger until we do so. If you don't disagree, can you rename the function? Thanks!

File: packages/ERTP/src/paymentLedger.js
Comment (dtribble) [inline packages/ERTP/src/paymentLedger.js:34]: It seems clunky to have to define this here, but it's non-blocking.

**Review by dtribble — APPROVED** [review]

File: packages/ERTP/src/paymentLedger.js
Comment (erights) [reply packages/ERTP/src/paymentLedger.js:34]: Ok. I'll merge as is but am interested in hearing suggestions

**Pull request merged**
